tsp-output/@azure-tools/typespec-csharp/src/Generated/Internal/ClientPipelineExtensions.cs

```diff
@@ -1,49 +1,36 @@
-// <auto-generated/>
 
-#nullable disable
+#nullable enable
 
-using System;
 using System.ClientModel;
 using System.ClientModel.Primitives;
-using System.Threading;
 using System.Threading.Tasks;
 
 namespace OpenAI
 {
     internal static class ClientPipelineExtensions
     {
-        public static async ValueTask<PipelineResponse> ProcessMessageAsync(this ClientPipeline pipeline, PipelineMessage message, RequestOptions requestContext, CancellationToken cancellationToken = default)
+        public static async ValueTask<PipelineResponse> ProcessMessageAsync(this ClientPipeline pipeline, PipelineMessage message, RequestOptions options)
         {
             await pipeline.SendAsync(message).ConfigureAwait(false);
 
-            if (message.Response == null)
+            if (message.Response!.IsError && options.ErrorOptions == ClientErrorBehaviors.Default)
             {
-                throw new InvalidOperationException("Failed to receive Result.");
+                throw await ClientResultException.CreateAsync(message.Response).ConfigureAwait(false);
             }
 
-            if (!message.Response.IsError || requestContext?.ErrorOptions == ClientErrorBehaviors.NoThrow)
-            {
-                return message.Response;
-            }
-
-            throw new ClientResultException(message.Response);
+            return message.Response;
         }
 
-        public static PipelineResponse ProcessMessage(this ClientPipeline pipeline, PipelineMessage message, RequestOptions requestContext, CancellationToken cancellationToken = default)
+        public static PipelineResponse ProcessMessage(this ClientPipeline pipeline, PipelineMessage message, RequestOptions options)
         {
             pipeline.Send(message);
 
-            if (message.Response == null)
-            {
-                throw new InvalidOperationException("Failed to receive Result.");
-            }
-
-            if (!message.Response.IsError || requestContext?.ErrorOptions == ClientErrorBehaviors.NoThrow)
+            if (message.Response!.IsError && options.ErrorOptions == ClientErrorBehaviors.Default)
             {
-                return message.Response;
+                throw new ClientResultException(message.Response);
             }
 
-            throw new ClientResultException(message.Response);
+            return message.Response;
         }
 
         public static async ValueTask<ClientResult<bool>> ProcessHeadAsBoolMessageAsync(this ClientPipeline pipeline, PipelineMessage message, RequestOptions requestContext)
```

Comment:
I don't think this is the correct logic for unbranded clients.  We don't have to revisit or block on this today for OpenAI, but we should address it soon so we don't ship an unbranded client with the wrong behavior.

One specific problem here is that we don't honor any response classifiers an end-user might pass.  There may be others as well, e.g. I don't think we want to return exploding responses from unbranded client -- it is not idiomatic to .NET to throw from properties on types.

Reply:
Tracking with https://github.com/Azure/autorest.csharp/issues/4479